lib/matplotlib/colors.py

```diff
@@ -1473,34 +1473,53 @@ def __init__(self, boundaries, ncolors, clip=False, *, extend='neither'):
         self.Ncmap = ncolors
         self.extend = extend
 
-        self._N = self.N - 1  # number of colors needed
+        self._n_regions = self.N - 1  # number of colors needed
         self._offset = 0
         if extend in ('min', 'both'):
-            self._N += 1
+            self._n_regions += 1
             self._offset = 1
         if extend in ('max', 'both'):
-            self._N += 1
-        if self._N > self.Ncmap:
-            raise ValueError(f"There are {self._N} color bins including "
-                             f"extensions, but ncolors = {ncolors}; "
-                             "ncolors must equal or exceed the number of "
-                             "bins")
+            self._n_regions += 1
+        if self._n_regions > self.Ncmap:
+            raise ValueError(f"There are {self._n_regions} color bins "
+                             "including extensions, but ncolors = "
+                             f"{ncolors}; ncolors must equal or exceed the "
+                             "number of bins")
 
     def __call__(self, value, clip=None):
         if clip is None:
             clip = self.clip
 
         xx, is_scalar = self.process_value(value)
         mask = np.ma.getmaskarray(xx)
+        # Fill masked values a value above the upper boundary
         xx = np.atleast_1d(xx.filled(self.vmax + 1))
         if clip:
             np.clip(xx, self.vmin, self.vmax, out=xx)
             max_col = self.Ncmap - 1
         else:
             max_col = self.Ncmap
+        # this gives us the bins in the lookup table in the range
+        # [0, _n_regions - 1]  (the offset is baked in in the init)
         iret = np.digitize(xx, self.boundaries) - 1 + self._offset
-        if self.Ncmap > self._N:
-            scalefac = (self.Ncmap - 1) / (self._N - 1)
+        # if we have more colors than regions, stretch the region index
+        # computed above to full range of the color bins.  This will
+        # make use skip some of the colors in the middle of the
+        # colormap but will use the full range
+        if self.Ncmap > self._n_regions:
+            # the maximum possible value in iret
+            divsor = self._n_regions - 1
+            if divsor == 0:
+                # special case the 1 region case.  What we should do here
+                # is a bit undefined (as _any_ color in the colormap is
+                # justifiable) we go with not adjusting the value (which will
+                # either be 0 or 1 depending on if we are extending down)
```

Comment:
I think this comment is also confusing.  The divisor will be 0 only if there are no extensions and 2 boundaries, or one extension and only 1 boundary.  So we have to pick one color out of the map.  BoundaryNorm is supposed to work like contourf, so we should do what contourf does, which is pick the middle of the colormap.  But we can't do that with scalefac, because iret is filled with zeros.  It has to be directly filled with (self.Ncmap -1) // 2.